js/upload/src/utils.ts

```diff
@@ -57,6 +57,25 @@ export function normalise_file(
 	return file;
 }
 
+function is_url(str: string): boolean {
+	try {
+	  const url = new URL(str);
+	  return url.protocol === "http:" || url.protocol === "https:";
+	} catch {
+	  return false;
+	}
+  }
+  
+export function get_fetchable_url_or_file(path: string | null, root: string, root_url: string | null): string {
```

Comment:
We should call this in `normalise_file` so that both are in synch if we ever need to make changes. Otherwise looks good to me! Agree we should be setting the url in the front end.